### PR TITLE
update Node admin doc

### PR DESCRIPTION
**Pull request by gmarek**
Fix kubernetes/kubernetes#30762

cc @wojtek-t

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/1136)

### CODE REVIEW

**Comment by jaredbhatti** [comment]
From a docs standpoint, this looks good. I have a question, which is should we explicitly state kubernetes versions in the docs, or should we assume that the user is using the latest version.  For example, should the sentence "In 1.4 release we updated the logic of node controller to better handle cases when a big number of Nodes have problems with reaching the master machine.."  I think we should reword that sentence to something like "since the 1.4 release" so that sentence still works post-1.4.

**Comment by gmarek** [comment]
I'm very far from being native speaker, but "Since 1.4 release we updated the logic" sounds strange to me.

@davidopp can you take a look at this change?

File: docs/admin/node.md [hunk 1]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
-status to ConditionUnknown when a node becomes unreachable (e.g. due to the node being down),
-and then later evicting all the pods from the node (using graceful termination) if the node
-continues to be unreachable. (The current timeouts for those are 40s and 5m, respectively.)
-It also allocates CIDR blocks to the new nodes.
+objects.
+
+Node controller has mutliple roles in Node's life. First is assigning a CIDR block to
+the Node when it is registered (if CIDR assignment is turned on). Second part is monitoring
+Node's health. Node controller is responsible for updating the NodeReady condition of node
+status to ConditionUnknown when a node becomes unreachable (i.e. Kubelet stopped sending heartbeats
```
Comment (davidopp) [inline docs/admin/node.md]: s/node status/NodeStatus/
s/i.e. Kubelet....being down/i.e. node controller stops receiving heartbeats e.g. due to the node being down/

File: docs/admin/node.md [hunk 2]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+e.g. due to the node being down), and then later evicting all the pods from the node
+(using graceful termination) if the node continues to be unreachable (the current timeouts
+for those are 40s and 5m, respectively). Node controller checks the state of each node every
```
Comment (davidopp) [inline docs/admin/node.md]: s/for those are 40s and 5m, respectively/are 40s to start reporting ConditionUnknown and 5m after that to start evicting pods/

And I would delete the last sentence ("Node controller checks..."). I think this is too much detail, and it's not clear what it means to "check the state" of a node without further explanation, since it's push based not pull based.

File: docs/admin/node.md [hunk 3]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+`--node-monitor-period` seconds.
+
+In 1.4 release we updated the logic of node controller to better handle cases when a
+big number of Nodes have problems with reaching the master machine (e.g. because
+master machine has networking problem). Since 1.4 node controller will look at the
```
Comment (davidopp) [inline docs/admin/node.md]: s/Since/Starting with/

I would change the "In previous releases..." sentence a bit; I think it's a little clearer to say "In previous releases if the master machine got partitioned from the cluster for at least 5m40s, it would delete all running pods from its "desired state" for the cluster, and then when the partition ended and the kubelets were able to contact the master again, they would see all of their pods had been deleted and would kill them."

File: docs/admin/node.md [hunk 4]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+releases if master machine got isolated from the cluster it would evict all running pods
+and after it would rejoin the cluster nodes would kill everything that was running.
+Since 1.4 this shouldn't happen.
+
+In normal circumstances (read: other than ones described in following paragraphs)
+node controller limits the eviction rate to `--node-eviction-rate` (default 0.1)
+per second, i.e. it won't evict pods from more than 1 node per 10 seconds.
+
+Nodes in the cluster are split into availablity zones. When a node in a given zone
+becomes unhealty node controller checks what percentage of nodes in this zone are
```
Comment (davidopp) [inline docs/admin/node.md]: s/unhealthy/unhealthy (NodeReady condition becomes ConditionUnknown or ConditionFalse)/

File: docs/admin/node.md [hunk 5]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+unhealthy in the same time. If the percentage is bigger than `--unhealthy-zone-threshold`
```
Comment (davidopp) [inline docs/admin/node.md]: s/in/at/

File: docs/admin/node.md [hunk 6]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+state of all nodes in the cluster when making a decision about pod eviction. In previous
+releases if master machine got isolated from the cluster it would evict all running pods
+and after it would rejoin the cluster nodes would kill everything that was running.
+Since 1.4 this shouldn't happen.
+
+In normal circumstances (read: other than ones described in following paragraphs)
+node controller limits the eviction rate to `--node-eviction-rate` (default 0.1)
+per second, i.e. it won't evict pods from more than 1 node per 10 seconds.
+
+Nodes in the cluster are split into availablity zones. When a node in a given zone
```
Comment (davidopp) [inline docs/admin/node.md]: I thought a bit about how to introduce the zone concept most compactly. I think the best thing to do is to remove the first sentence and change the second one to "When a node in a given availability zone (note: if your cluster does not span multiple cloud provider availability zones, then there is only one availability zone, namely the whole cluster) becomes unhealthy, node controller checks what percentage of nodes in the zone are unhealthy (NodeReady condition becomes ConditionUnknown or ConditionFalse) at the same time."

File: docs/admin/node.md (hunk 5 above)
Comment (davidopp) [inline docs/admin/node.md]: I believe it's "at least" not "bigger than" ?
Also, it's expressed as a fraction not a percentage, so 0.55

File: docs/admin/node.md [hunk 7]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+becomes unhealty node controller checks what percentage of nodes in this zone are
+unhealthy in the same time. If the percentage is bigger than `--unhealthy-zone-threshold`
+(default 55) then the eviction rate is reduced: if the cluster is small (i.e. has less than
```
Comment (davidopp) [inline docs/admin/node.md]: I believe it's "less than or equal to"

File: docs/admin/node.md [hunk 8]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+`--large-cluster-size-threshold` nodes - default 50) then evictions are stopped, if the
```
Comment (davidopp) [inline docs/admin/node.md]: s/if the cluster is big/otherwise/

File: docs/admin/node.md [hunk 9]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+cluster is big the eviction rate is reduced to `--secondary-node-eviction-rate` (default 0.01)
+per second.
```
Comment (davidopp) [inline docs/admin/node.md]: I would add a sentence at the end like "The reason these policies are implemented per availability zone is because one availability zone might become partitioned from the master while the others remain connected."

(BTW, we should think about how this should work when we have HA master with one master per zone. Do the same thing? Or treat it as a single cluster with no zones? Or some other policy)

File: docs/admin/node.md [hunk 10]
```diff
@@ -120,25 +120,48 @@ Currently, there are three components that interact with the Kubernetes node int
...
+
+To allow correct failover to another zone when one is completely down, as this is why we can
```
Comment (davidopp) [inline docs/admin/node.md]: I would replace the first sentence with: "A key reason for spreading your nodes across availability zones is so that workload can be shifted to healthy zones when one entire zone goes down. To enable this behavior, if all nodes in a zone are unhealthy then node controller evicts at the normal rate `--node-eviction-rate`."

File: docs/admin/node.md (hunk 1 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 2 above)
Comment (gmarek) [inline docs/admin/node.md]: I need to mention `--node-monitor-period` flag somewhere - I added this sentence only because of that.

File: docs/admin/node.md (hunk 3 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 6 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 5 above)
Comment (gmarek) [inline docs/admin/node.md]: Yeah, you're right.

File: docs/admin/node.md (hunk 7 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 9 above)
Comment (gmarek) [inline docs/admin/node.md]: Good question. We didn't thought about that - we probably should figure it out before 1.5. I think we'll need to take into the account if the master in the given availability zone is in the quorum or not. This is a longer discussion.

File: docs/admin/node.md (hunk 10 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

**Comment by gmarek** [comment]
Thanks - can you PTAL @davidopp

File: docs/admin/node.md [hunk 11]
```diff
@@ -120,25 +120,52 @@ Currently, there are three components that interact with the Kubernetes node int
...
+Node controller checks the state of each node every `--node-monitor-period` seconds.
+
+In 1.4 release we updated the logic of node controller to better handle cases when a
+big number of Nodes have problems with reaching the master machine (e.g. because
+master machine has networking problem). Starting with 1.4 node controller will look at the
+state of all nodes in the cluster when making a decision about pod eviction. In previous
+releases if the master machine got partitioned from the cluster for at least 5m40s,
+it would delete all running pods from its "desired state" for the cluster,
+and then when the partition ended and the kubelets were able to contact the master again,
+they would see all of their pods had been deleted and would kill them. Since 1.4 this shouldn't happen.
```
Comment (davidopp) [inline docs/admin/node.md]: I would say "Starting in 1.4 this shouldn't happen, and instead we use the policy described below."

File: docs/admin/node.md [hunk 12]
```diff
@@ -120,25 +120,52 @@ Currently, there are three components that interact with the Kubernetes node int
...
+
+In normal circumstances (read: other than ones described in following paragraphs)
+node controller limits the eviction rate to `--node-eviction-rate` (default 0.1)
+per second, i.e. it won't evict pods from more than 1 node per 10 seconds.
+
+When a node in a given availability zone (note: if your cluster does not span multiple cloud
+provider availability zones, then there is only one availability zone, namely the whole cluster)
+becomes unhealthy, node controller checks what percentage of nodes in the zone are unhealthy
+(NodeReady condition becomes ConditionUnknown or ConditionFalse) at the same time. If the
```
Comment (davidopp) [inline docs/admin/node.md]: s/becomes/is/

Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md [hunk 13]
```diff
@@ -120,25 +120,52 @@ Currently, there are three components that interact with the Kubernetes node int
...
+zone might become partitioned from the master while the others remain connected.
+
+A key reason for spreading your nodes across availability zones is so that workload can be
+shifted to healthy zones when one entire zone goes down. To enable this behavior, if all
+nodes in a zone are unhealthy then node controller evicts at the normal rate `--node-eviction-rate`.
+The corner case for that is when all zones are completely unhealthy (i.e. there's no healthy node in
+the cluster). In such case node controller assumes that there's some problem with master machine
+connectivity and stops all evictions until any connectivity is restored.
+
+Third responsibility of node controler is keeping the list of nodes up to date with
```
Comment (davidopp) [inline docs/admin/node.md]: I would rephrase this as "Third responsibility of node controller is keeping the node controller's list of nodes up to date with the cloud provider's list of available machines."

And actually I would make this second, and move the eviction stuff to third (so it's the last part of this section).

File: docs/admin/node.md [hunk 14]
```diff
@@ -120,25 +120,52 @@ Currently, there are three components that interact with the Kubernetes node int
...
+the list of avialable machines. When running in cloud environment whenever the node is
```
Comment (davidopp) [inline docs/admin/node.md]: s/the node/a node/

File: docs/admin/node.md (hunk 13 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 14 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md [hunk 15]
```diff
@@ -120,25 +120,53 @@ Currently, there are three components that interact with the Kubernetes node int
...
-and then later evicting all the pods from the node (using graceful termination) if the node
-continues to be unreachable. (The current timeouts for those are 40s and 5m, respectively.)
-It also allocates CIDR blocks to the new nodes.
+objects.
+
+Node controller has mutliple roles in Node's life. First is assigning a CIDR block to
+the Node when it is registered (if CIDR assignment is turned on). Second is keeping the
+node controller's list of nodes up to date with the cloud provider's list of available
+machines. When running in cloud environment whenever a node is unhealthy node controller
+asks cloud provider if the VM for that node is still available. If not it deletes the node.
```
Comment (davidopp) [inline docs/admin/node.md]: I would change that last sentence to "If not, the node controller deletes the node from its list of nodes."

**Comment by davidopp** [comment]
LGTM after you make the one last comment I had.

File: docs/admin/node.md (hunk 15 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

**Comment by davidopp** [comment]
LGTM

**Comment by gmarek** [comment]
Squashed commits. @jaredbhatti PTAL

**Comment by gmarek** [comment]
Ping @jaredbhatti

**Comment by gmarek** [comment]
Closing in favor of #1195
